design-proposals/2023-11-08-complex-animatable-interactive-annotations.md

```diff
@@ -0,0 +1,55 @@
+
+An annotation as described in this proposal is a piece of UI that is drawn on the map at a single coordinate. In Apple MapKit this would be an MKAnnotationView, and in Google world this would be similar to a Marker. Annotations are expected to be user interactable, animatable and allow for complex ui to be represented.
+
+## Motivation
+
+Most mobile map experiences display some sort of custom content on the map, often in the form of annotations. MapLibre’s current annotation system is difficult to use, performs poorly, and lacks many of the features developers expect from modern mapping libraries.
+
+This proposal introduces the concept annotations that are complex, animatable and interactive to maplibre’s core. The first phase allows bitmap backed annotations to be rendered. The second phase will enable native platform views (ios, android) to be rendered directly by maplibre-gl.
+
+These new annotations will provide a more flexible tool for drawing user content on the map that is more inline with what developers expect from the built-in mobile map toolkits. The focus on animations and interactivity allows developers to create differentiated map experiences that feel made for mobile.
+
+By introducing a specialized rendering path for annotations, we bypass some of the limitations of the current layer-based system. This leads to better performance, especially in scenarios with many dynamic elements, where full layout re-calculations would otherwise cause noticeable lag and degrade the user experience.
+
+The customization options that these annotations will provide, together with the existing MapLibre stack, aim to offer a system that is more adaptable and robust than those from Google or Apple.
+
+## Proposed Change
+
+### Phase 1
+For this phase we would like to just build our support for bitmaps. The idea being that all native views eventually become bitmaps and supporting them robustly will eventually allow the renderer to draw native views correctly.
+
+* Introduce “Annotation” as a first class citizen of maplibre-gl core c++ layer. This feature would only be available to native clients and platforms and not maplibre-js.
+* Annotations rely on their own bitmap handling and not on the symbol layer texture atlas. Bitmaps should be able to be transformed and quickly swappable to support animations.
+* Annotations can define a label (reuse symbol layer logic)
+* Annotations provide callbacks and handling / can trigger full map re-renders. These callbacks need to be in sync with the native render loop so that animations can be synchronized properly.
+* Annotations participate in label and symbol layer collisions.
+[Stretch] Annotations support clustering, like symbol layers do.
+* Platform layers allow annotations to participate in hit tests and “onTap” events.
+* Layout, styling and positioning of Annotations can happen outside of the style’s layout and paint loops. However they should still participate in the rendering process.
```

Comment:
This seems to conflict with the earlier point that annotations participate in label and symbol layer collisions.

Reply:
There is a balancing act here where we don't want every frame of an Annotation animation to trigger a global map re-layout and re-styling. But once animations "settle" (or through some other criteria, maybe a debounce) the annotation should once again participate in collisions etc.

For example, if we have an animation that moves from one point of the map to another. The movement animation needs to be smooth (maybe even at the expense of any layout which may happen underneath it). If while it move it collides with a map label (and we determine the label should be hidden), then we want to gracefully fade out that label on the next layout pass.